packages/orm/src/relations/hasPersistedAsyncRelation.ts

```diff
@@ -0,0 +1,342 @@
+    hint,
+    fn
+  );
+}
+
+export class PersistedAsyncRelationImpl<T extends Entity, U extends Entity, H extends ReactiveHint<T>, N extends never | undefined>
+  extends AbstractRelationImpl<U>
+  implements PersistedAsyncRelation<T, U, N>
+{
+  readonly #entity: T;
+  readonly #fieldName: keyof T & string;
+  readonly #reactiveHint: Const<H>;
+  // Either the loaded entity, or N/undefined if we're allowed to be null
+  private loaded!: U | N | undefined;
+  // We need a separate boolean to b/c loaded == undefined can still mean "_isLoaded" for nullable fks.
+  private _isLoaded = false;
+  private loadPromise: any;
+  constructor(
+    entity: T,
+    otherMeta: EntityMetadata<U>,
+    private fieldName: keyof T & string,
+    public otherFieldName: keyof U & string,
+    public reactiveHint: Const<H>,
+    private fn: (entity: Reacted<T, H>) => U | N,
+  ) {
+    super();
+    this.#entity = entity;
+    this.#fieldName = fieldName;
+    this.#reactiveHint = reactiveHint;
+  }
+
+  async load(opts?: { withDeleted?: true, forceReload?: true }): Promise<U | N> {
+    ensureNotDeleted(this.#entity, "pending");
+    const { loadHint } = this;
+    if (!this._isLoaded) {
+      return (this.loadPromise ??= this.#entity.em.populate(this.#entity, loadHint).then(() => {
+        this._isLoaded = true;
+        // Go through `this.get` so that `setField` is called to set our latest value
+        return this.doGet(opts);
+      }));
+    }
+    return Promise.resolve(this.doGet(opts));
+  }
+
+  private doGet(opts?: { withDeleted?: boolean }): U | N {
+    const { fn } = this;
+    ensureNotDeleted(this.#entity, "pending");
+    if (this._isLoaded || (!this.isSet && isLoaded(this.#entity, this.loadHint))) {
+      const newValue = this.filterDeleted(fn(this.#entity as Reacted<T, H>), opts);
+      // It's cheap to set this every time we're called, i.e. even if it's not the
+      // official "being called during em.flush" update (...unless we're accessing it
+      // during the validate phase of `em.flush`, then skip it to avoid tripping up
+      // the "cannot change entities during flush" logic.)
+      if (!(this.#entity.em as any)._isValidating) {
+        this.set(newValue);
+      }
+      return this.maybeFindEntity();
+    } else if (this.isSet) {
+      return this.#entity.__orm.data[this.fieldName];
+    } else {
+      throw new Error(`${this.fieldName} has not been derived yet`);
+    }
+  }
+
+  get fieldValue(): U {
+    return this.#entity.__orm.data[this.fieldName];
+  }
+
+  get getWithDeleted(): U | N {
+    return this.doGet({ withDeleted: true });
+  }
+
+  get get(): U | N {
+    return this.doGet({ withDeleted: false });
+  }
+
+  get isLoaded(): boolean {
+    return this._isLoaded;
+  }
+
+  set(other: U | N): void {
+    this.setImpl(other);
+  }
+
+  get isSet(): boolean {
+    return this.current() !== undefined;
+  }
+
+  // Internal method used by OneToManyCollection
+  setImpl(other: U | IdOf<U> | N): void {
```

Comment:
Musing: Seems like callers doing `o2m.add/remove` should be unsupported; I wonder if eventually we have a `ReadOnlyCollection` / `DerivedOneToManyCollection` that does not have `add` / `remove` methods.

Reply:
What happens if:

1. `Author.favoriteBook = b1`
2. Someone loads `b1.favoritedAuthors.load()`, returns `[a1]`
3. Someone creates `b2` with a higher rating
   * There are no obvious setter/add calls for us to know to update `b1.favoritedAuthors`...
5. `a1.favoriteBook.get` will immediately return `b2` (great)
6. Ideally `b1.favoritedAuthors.get` would now return `[]`

We don't need to solve this for this PR, but the ergonomics of "always up-to-date both sides of relations" is an important feature of joist.